modules/rostests/kmtests/ntos_io/IoDeviceInterface.c

```diff
@@ -1,125 +1,214 @@
-    DEVICE_OBJECT DeviceObject;
-    EXTENDED_DEVOBJ_EXTENSION DeviceObjectExtension;
-    DEVICE_NODE DeviceNode;
     UNICODE_STRING SymbolicLinkName;
-    NTSTATUS Status;
 
-    RtlInitUnicodeString(&SymbolicLinkName, L"");
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        UNICODE_STRING AliasSymbolicLinkName;
+        NTSTATUS Status = IoGetDeviceInterfaceAlias(&SymbolicLinkName, Types[n], &AliasSymbolicLinkName);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoGetDeviceInterfaceAlias(): success: %d %wZ\n", n, &AliasSymbolicLinkName);
+
+            RtlFreeUnicodeString(&AliasSymbolicLinkName);
+        }
+        else
+        {
+            trace("IoGetDeviceInterfaceAlias(): fail: %d 0x%x\n", n, Status);
+        }
+    }
+}
+
+static
+VOID
+Test_IoSetDeviceInterfaceState(PWSTR SymbolicLink)
+{
+    UNICODE_STRING SymbolicLinkName;
+
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        NTSTATUS Status = IoSetDeviceInterfaceState(&SymbolicLinkName, TRUE);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoSetDeviceInterfaceState(): successfully enabled interface: %d %wZ\n", n, &SymbolicLinkName);
+        }
+        else
+        {
+            trace("IoSetDeviceInterfaceState(): failed to enable interface: %d 0x%x\n", n, Status);
+        }
+    }
+}
+
+static
+VOID
+Test_IoGetDeviceInterfaces(const GUID* guid)
+{
+    NTSTATUS Status;
+    PWSTR SymbolicLinkList, SymbolicLinkListPtr;
+    UNICODE_STRING GuidString;
 
-    // Prepare our surrogate of a Device Object
-    DeviceObject.DeviceObjectExtension = (PDEVOBJ_EXTENSION)&DeviceObjectExtension;
+    RtlStringFromGUID(guid, &GuidString);
 
-    // 1. DeviceNode = NULL
-    DeviceObjectExtension.DeviceNode = NULL;
-    Status = IoRegisterDeviceInterface(&DeviceObject, &Guid, NULL,
-        &SymbolicLinkName);
+    Status = IoGetDeviceInterfaces(guid, NULL, DEVICE_INTERFACE_INCLUDE_NONACTIVE, &SymbolicLinkList);
 
-    ok(Status == STATUS_INVALID_DEVICE_REQUEST,
-        "IoRegisterDeviceInterface returned 0x%08lX\n", Status);
+    ok(NT_SUCCESS(Status), "IoGetDeviceInterfaces failed with status 0x%X for '%wZ'\n", Status, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        return;
+    }
 
-    // 2. DeviceNode->InstancePath is of a null length
-    DeviceObjectExtension.DeviceNode = &DeviceNode;
-    DeviceNode.InstancePath.Length = 0;
-    Status = IoRegisterDeviceInterface(&DeviceObject, &Guid, NULL,
-        &SymbolicLinkName);
+    trace("IoGetDeviceInterfaces '%wZ' results:\n", &GuidString);
+    RtlFreeUnicodeString(&GuidString);
 
-    ok(Status == STATUS_INVALID_DEVICE_REQUEST,
-        "IoRegisterDeviceInterface returned 0x%08lX\n", Status);
+    for (SymbolicLinkListPtr = SymbolicLinkList;
+        SymbolicLinkListPtr[0] != 0 && SymbolicLinkListPtr[1] != 0;
```

Comment:
The `SymbolicLinkListPtr[1] != 0` here would make you skip any strings that are only one character long.
While that's not an expected outcome here, it seems unnecessary to provide such an arbitrary limitation.

Reply:
Removed.